packages/flutter_tools/lib/src/vmservice.dart

```diff
@@ -218,82 +203,63 @@ class VMService {
-        'service': 'hotRestart',
-        'alias': 'Flutter Tools',
-      });
+      _delegateService.registerService('hotRestart', 'Flutter Tools');
     }
 
-    _peer.registerMethod('flutterVersion', (rpc.Parameters params) async {
+    _delegateService.registerServiceCallback('flutterVersion', (Map<String, dynamic> params) async {
       final FlutterVersion version = context.get<FlutterVersion>() ?? FlutterVersion();
       final Map<String, Object> versionJson = version.toJson();
       versionJson['frameworkRevisionShort'] = version.frameworkRevisionShort;
       versionJson['engineRevisionShort'] = version.engineRevisionShort;
       return versionJson;
     });
-
-    _peer.sendNotification('registerService', <String, String>{
-      'service': 'flutterVersion',
-      'alias': 'Flutter Tools',
-    });
+    _delegateService.registerService('flutterVersion', 'Flutter Tools');
 
     if (compileExpression != null) {
-      _peer.registerMethod('compileExpression', (rpc.Parameters params) async {
-        final String isolateId = params['isolateId'].asString;
+      _delegateService.registerServiceCallback('compileExpression', (Map<String, dynamic> params) async {
+        final String isolateId = params['isolateId'] as String;
         if (isolateId is! String || isolateId.isEmpty) {
           throw rpc.RpcException.invalidParams(
               "Invalid 'isolateId': $isolateId");
         }
-        final String expression = params['expression'].asString;
+        final String expression = params['expression'] as String;
         if (expression is! String || expression.isEmpty) {
           throw rpc.RpcException.invalidParams(
               "Invalid 'expression': $expression");
         }
-        final List<String> definitions =
-            List<String>.from(params['definitions'].asList);
-        final List<String> typeDefinitions =
-            List<String>.from(params['typeDefinitions'].asList);
-        final String libraryUri = params['libraryUri'].asString;
-        final String klass = params['klass'].exists ? params['klass'].asString : null;
-        final bool isStatic = params['isStatic'].asBoolOr(false);
-
+        final List<String> definitions = List<String>.from(params['definitions'] as List<dynamic>);
+        final List<String> typeDefinitions = List<String>.from(params['typeDefinitions'] as List<dynamic>);
+        final String libraryUri = params['libraryUri'] as String;
+        final String klass = params['klass'] as String;
+        final bool isStatic = params['isStatic'] as bool ?? false;
         try {
           final String kernelBytesBase64 = await compileExpression(isolateId,
               expression, definitions, typeDefinitions, libraryUri, klass,
               isStatic);
-          return <String, dynamic>{'type': 'Success',
-            'result': <String, dynamic> {'kernelBytes': kernelBytesBase64}};
+          return <String, dynamic>{
```

Comment:
This was required to make the tests pass, but is incredibly confusing. the vm essentially expects this double result structure and I can't figure out why.

Reply:
fyi @bkonyi

Reply:
Hm, that is odd. [Here's where we do that access in the service](https://github.com/dart-lang/sdk/blob/master/sdk/lib/vmservice/running_isolates.dart#L161), but I'm not sure why we do it... @aam do you happen to know why we expect this format from external compilation services?

Reply:
Not sure about the question, but redgarding two `'result'`s in `jsonMap`: one comes from the vm itself, second comes from `message.forwardToJson` that wraps vm response into `result`
https://github.com/dart-lang/sdk/blob/master/sdk/lib/vmservice/message.dart#L131